csrc/ir_nodes.cpp

```diff
@@ -2500,6 +2500,33 @@ void IterDomain::parallelize(ParallelType t) {
     return;
   }
 
+  // Leaf domains are domains that are not used by any other domains.
+  auto isLeafDomain = [&]() {
+    for (auto expr : uses()) {
+      if (expr->isA<Split>() || expr->isA<Merge>() || expr->isA<Resize>()) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  // assert check that we only parallelize a leaf domain.
+  // this check is only applied to isParallelTypeThread
+  // getNonGlobalProducerStridedIndices in NVFuserTest.FusionIndexHoist2_CUDA
+  // is trying to parallelize a non-leaf domain using vectorize
```

Comment:
This is interesting. Do you know why?

Reply:
The error is due to the following code:
```
  c2p_index_map = replay_producer_as_consumer.getReplay();
  p2c_index_map = invertOneToOneMap(c2p_index_map);

  // Forward vectorized IDs to index into producer correctly
  // We want p_id to be vectorized like consumer just for the indexing, then we
  // need to switch it back later. Store previous state here when changing. We
  // need to do this as replaying producer as consumer can use replay best
  // effort which means some domains may be the originals.
  std::vector<std::pair<IterDomain*, ParallelType>> p_id_backup;
  for (auto entry : c2p_index_map) {
    auto ref_id = GpuLower::current()->caMap()->getConcreteMappedID(
        entry.first, IdMappingMode::EXACT);
    auto p_id = entry.second;
    if (ref_id->getParallelType() == ParallelType::Vectorize) {
      p_id_backup.emplace_back(p_id, p_id->getParallelType());
      p_id->parallelize(ParallelType::Vectorize);
    } else if (ref_id->getParallelType() == ParallelType::MisalignedVectorize) {
      p_id->parallelize(ParallelType::MisalignedVectorize);
    }
  }
```
It is trying to parallelize` iV15{4}, iV11{4}, and iS9{4} `using ref_id: `iV7{4}.` Here` iS9{4}` is not a leaf domain.

Reply:
What does the producer domain look like?

Reply:
producer_tv= `T4_l[ iS8{( ceilDiv(T0.size[0], 4) )}, iS9{4} ] ca_pos( 1 ) produce_pos( 1 )`
consumer_tv= T5_g[ iS6{( ceilDiv(T0.size[0], 4) )}, iV7{4} ] produce_pos( 1 )
unknown file: Failure
C++ exception with description "Only allowed to parallelize a leaf domain. Domain: `iS9{4}`, Parallel type: V, Definition: Split: iS24{T0.size[0]} by factor 4 -> iS8{( ceilDiv(T0.size[0], 4) )}, iS9{4}, start offset: 0, stop offset: 0

But T4_l is actually:
```
T4_l[ iS8{( ceilDiv(i0, 4) )}, iS18{( ceilDiv(4, 3) )}, iS19{3} ] ca_pos( 1 ) produce_pos( 1 )
 root domain : (iS4{i0})
 contiguity: t
  Split: iS4{i0} by factor 4 -> iS8{( ceilDiv(i0, 4) )}, iS9{4}, start offset: 0, stop offset: 0
  Split: iS9{4} by factor 3 -> iS18{( ceilDiv(4, 3) )}, iS19{3}, start offset: 0, stop offset: 0
 leaf domain : (iS8{( ceilDiv(i0, 4) )}, iS18{( ceilDiv(4, 3) )}, iS19{3}) 
```
This is because 
```
  ir_utils::TVDomainGuard domain_guard(
      producer_tv, producer_replayed_as_consumer);
```
where we want to view producer T4 same to its consumer T5. The view and vectorization is just for indexing purpose and after that the parallelization is reversed.

Reply:
Hmm, I wonder this vectorization propagation is still necessary. Can you try commenting the code out?

Reply:
I deleted the " vectorization propagation " part and all the tests passed.

Reply:
Thanks for checking. Let me double-check if this can really be removed.

Reply:
Since it's a different issue, I created a separate PR (#267).